### ISSUE TITLE
Nurbs Trajectory

### ISSUE THREAD

**Issue by aktasalifurkan** [esmini/esmini#144]
Hello,

Currently, I am working on trajectories however when I used the NURBS trajectory, Ego vehicle gets new a heading angle instantly and at the end, the heading angle is not how it is supposed to be.

```
<FollowTrajectoryAction>
    <Trajectory closed="False" name="TrajectoryNurbs">
        <ParameterDeclarations/>
        <Shape>
            <Nurbs order="3">
                <ControlPoint weight="1">
                    <Position>
                        <RelativeObjectPosition entityRef="Ego" dx="0" dy="0" dz="0">
                            <Orientation h="0.0" p="0.0" r="0.0" type="relative"/>
                        </RelativeObjectPosition>/>
                    </Position>
                </ControlPoint>
                <ControlPoint weight="0.707106781186548">
                    <Position>
                        <RelativeObjectPosition entityRef="Ego" dx="9" dy="0" dz="0">
                            <Orientation h="0.0" p="0.0" r="0.0" type="relative"/>
                        </RelativeObjectPosition>
                    </Position>
                </ControlPoint>
                <ControlPoint weight="1.0">
                    <Position>
                        <RelativeObjectPosition entityRef="Ego" dx="9" dy="9" dz="0">
                            <Orientation h="0.0" p="0.0" r="0.0" type="relative"/>
                        </RelativeObjectPosition>
                    </Position>
                </ControlPoint>
                <Knot value="0.0" />
                <Knot value="0.0" />
                <Knot value="0.0" />
                <Knot value="1.0" />
                <Knot value="1.0" />
                <Knot value="1.0" />
            </Nurbs>
        </Shape>
    </Trajectory>
    <TimeReference>
        <None/>
    </TimeReference>
    <TrajectoryFollowingMode followingMode="position"/>
</FollowTrajectoryAction>
```

In this trajectory, I expect a quarter circle with a 9 m radius however it is not what I get. In the first step, It skips to 0.114430 rad (6.556°) and it finishes at 1.533015 rad (87.835°). Also, It skips to a new heading angle multiple times during the action. It will be appreciated if you guys fix this issue.

BR.

**Comment by eknabevcc** [comment]
You're right. There's two issues:

**Heading angle not perfectly correct at start and end of trajectory.**
The heading of a nurbs trajectory is calculated based on a polyline approximation of the nurbs curve. A side effect is that heading will not be perfect at start and end. BUT when investigating your issue we found a bug which made the initial heading error unnecessary big. That bug is now fixed, so heading error should now be reduced, however not eliminated.

**Heading changes in a non continuous manner**
This was another bug we found thanks to your report. The heading was not interpolated correctly in between (approximation) polyline segments. That is also fixed, resulting in smoother heading changes.

Commit 9d6a7ba32c7d3068af7a7d5164d8185585d39604 and [release 2.12.5](https://github.com/esmini/esmini/releases).

**Comment by aktasalifurkan** [comment]
Yes, I tried it and still, it has the problem. At start, it skips a certain angle again but it is better. However, at the end, I do not know why but it looks like the vehicle maintains its heading for 0.32 seconds. The trajectory contains two clothoid and one nurbs. I am adding that is expected as a picture and the openscenario file so that you can look into it.
[Traj.zip](https://github.com/esmini/esmini/files/6999961/Traj.zip)

**Comment by eknabevcc** [comment]
Thanks for input. Yes, there are (hopefully were) a few remaining issues:
1. The angle error comes from heading being based on polyline approximation of the nurbs. This was a lazy solution from our side. And we have now fixed it so that heading is based on actual nurbs curve tangent instead, and hence should be more correct. 
Fixed in commit c456239899cfed040c7550419e61cc2e294f1c84 and part of release [v2.12.6](https://github.com/esmini/esmini/releases/tag/v2.12.6).

2. Glitches between the trajectories. I see you put each FollowTrajectoryAction in separate Acts. For events that should start right after another its better to put in same maneuver, to avoid delay that comes with extra time step to trig act and it's maneuvers. This is probably a limitation in esmini, not to blame the OpenSCENARIO standard.

3. The 0.32s tail in the end. Yes, when the FollowTrajectoryAction is done, the default behavior is to maintain speed and heading relative to road and just move vehicle forward along the road keeping lane offset. I guess the scenario does not end exactly when the trajectory action ends. My suggestion is to add a "halt" event (set speed = 0) to make sure the vehicle stops when expected.

4. I like how you approximate an arc with a nurbs curve. In the specific case of arc it may be simpler to represent it as a clothoid with curvatureDot="0.0". But the nurbs seems to give very similar result.

2, 3 and 4 are demonstrated in attached scenario files: [Traj-mods.zip](https://github.com/esmini/esmini/files/7006126/Traj-mods.zip)

**Comment by johschmitz** [comment]
I am also having some trouble with trajectories. See the attached example with one NURBS and one Polyline trajectory:
[example.zip](https://github.com/esmini/esmini/files/7073160/example.zip)

Is the vehicle supposed to continue with a heading in lane direction when exiting the trajectory?
The NURBS trajectory seems to not work correctly but I don't see what is wrong.
How am I supposed to correctly define the NURBS trajectory timing? Maybe for that I need to open a second issue in the scenariogeneration repo(?)

**Comment by johschmitz** [comment]
@eknabevcc Oh dear, now I see I had an absolute path in the example ... good that we did this. Fixed it.

I don't see how to set TimeRef to None with scenariogeneration lib. Will open another ticket over there. On the other hand I don't see why esmini should not just ignore the TimeRef when no times are defined? Or maybe quit with an error? Also do you know where to find an explanation about how to add a time reference to a NURBS trajectory in OpenSCENARIO. Is this even possible? What is the mathematical definition and intuition?

Regarding the TeleportAction I tried to add it as an additional init actions but it does not work. Is this impossible? Do I need a complete full blown storyboard?

**Comment by aktasalifurkan** [comment]
The idea of using clothoid to create arc is something that did not come to my mind. Thanks to this solution, I can create the desired trajectory with 3 clothoid trajectories. Also, I do not need to calculate the NURBS control points for every trajectory change. It is a wonderful fix, well done.

**Comment by johschmitz** [comment]
I manged to set the time reference to None with scenariogeneration lib as explained in the other ticket. Released in the a new version of the Blender add-on https://github.com/johschmitz/blender-driving-scenario-creator/releases/tag/v0.5.0 which now has a NURBS trajectory drawing tool.

However, my other questions from above remain open.

**Comment by eknabevcc** [comment]
@aktasalifurkan Good to hear that it works better now.

@johschmitz I agree that esmini should ignore TimeRef whenever trajectory duration is 0 according to timestamps. Thanks for pointing out that issue. It's now fixed in commit cc6238ca1c0ade9aefb94a1e7f2e48bc143b8e1f.
BTW, I was wrong about esmini not supporting Nurbs time stamps. I simply forgot that was added :).

When it comes to the TeleportAction, the reason why you can't add it to the Init section is simply because it should be applied not initially but after the FollowTrajectoryAction has finished. Elaborating on heading topic:
In older versions esmini DefaultController always aligned entities to the road direction. However some emerging use cases required explicit entity orientation. So now esmini only align entities initially if and only if orientation is not set explicitly. Then it will keep same heading (relative to road direction). 
So after a FollowTrajectory, the default behavior is to keep same relative heading. If alignment is desired there are two ways: 1. Make sure trajectory heading do align with road at endpoint. 2. Add teleport action with relative heading = 0, as suggested above (if trajectory end heading differ from road direction this will result in a discreet change/jump of heading)

**Comment by eknabevcc** [comment]
Initial issue solved. And additional questions sorted out. 
Closing issue. Please re-open for further questions on the topics.
